File: agent-driven-architecture-main/examples/typescript/src/blocks/analysis/fold.ts
// ── blocks/analysis/fold — the block's ARM (12.4) ──────────────────────────
// The asymmetry between the two arms IS the untrusted-recall rule, written as
// code rather than as a warning:
//
//   recallAnalysis   appends a note. EMITS NO EFFECT AT ALL. Recalled content
//                    cannot reach the relay, let alone anything irreversible.
//   publishAnalysis  appends the conclusion AND emits `PublishConclusion` — the
//                    deep tier's own signed act, from its own success branch.

import type { Signature } from "@adr/spine/pure/actor";
import type { Timestamp } from "@adr/spine/pure/ids";
import type { ArmOut } from "@adr/spine/pure/verb";
import { armOut } from "@adr/spine/pure/verb";
import type { AnalysisResult, PublishConclusion } from "./contract";
import type { AnalysisSlice } from "./slice";
import { withNote, withPublished } from "./slice";

export function analysisArm(
  slice: AnalysisSlice,
  r: AnalysisResult,
  now: Timestamp,
  _sig: Signature,
): ArmOut<AnalysisSlice> {
  switch (r.tool) {
    case "recallAnalysis":
      // `now` is the ONE clock read the boundary made; pairing it with the
      // recall's own `publishedAt` is what makes the age replayable.
      return armOut(withNote(slice, { at: now, recall: r.recall }), [], []);
    case "publishAnalysis": {
      const effect: PublishConclusion = {
        kind: "PublishConclusion",
        at: now,
        effectClass: "Routine",
        text: r.text,
      };
      return armOut(withPublished(slice, r.text), [effect], []);
    }
    default: {
      const _never: never = r;
      return _never;
    }
  }
}

File: agent-driven-architecture-main/examples/kotlin/block/analysis/src/main/kotlin/adr/blocks/analysis/Fold.kt
// ── blocks/analysis/fold — the block's ARM ─────────────────────────────────
// Exhaustive over the block's sealed sub-union, with NO else arm. The same three
// rules every arm follows (§7): read current state before deciding, push effects
// only inside the success branch, and fold a per-item Notice on rejection.
//
// The asymmetry between the two arms IS the tiering rule, in code:
// recall APPENDS A NOTE AND EMITS NOTHING; publish is the only path to an effect.
// A conclusion the fast tier recalled therefore cannot be re-published, cannot page
// anyone, and cannot reach any irreversible act on its own (11.3).

package adr.blocks.analysis

import adr.contract.AnalysisEffect
import adr.contract.AnalysisResult
import adr.spine.pure.ArmOut
import adr.spine.pure.Notice
import adr.spine.pure.Signature
import adr.spine.pure.Timestamp

internal class AnalysisArm {

    fun arm(
        slice: AnalysisSlice,
        result: AnalysisResult,
        now: Timestamp,
        sig: Signature,
    ): ArmOut<AnalysisSlice> = when (result) {
        // The recall is recorded — including WHICH BRANCH it took — and nothing else.
        is AnalysisResult.RecallAnalysis ->
            ArmOut(slice = slice.withNote(AnalysisNote(now, result.recall)))

        is AnalysisResult.PublishAnalysis ->
            if (result.text.isBlank()) {
                ArmOut(
                    slice = slice,
                    notices = listOf(Notice.Rejected(now, result.tool, "a conclusion cannot be blank")),
                )
            } else {
                ArmOut(
                    slice = slice.withPublished(result.text),
                    effects = listOf(AnalysisEffect.PublishConclusion(now, result.text)),
                )
            }
    }
}
